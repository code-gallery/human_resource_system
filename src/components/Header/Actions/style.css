.border-btn {
  float: left;
  margin-top: 3px;
}
