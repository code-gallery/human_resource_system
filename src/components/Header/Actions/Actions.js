import React, { Component } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import './style.css'

class Actions extends Component {
  renderLoading() {
    return (
      <ul className="changes-btns">
        <li>
          <img src="/assets/loading.gif" alt="loading" />
        </li>
      </ul>
    )
  }

  renderEditAction() {
    const { toggleEditMode, editLabel } = this.props

    return (
      <ul className="changes-btns">
        <li>
          <button
            className="border-btn"
            onClick={toggleEditMode}
          >
            {editLabel}
          </button>
        </li>
      </ul>
    )
  }

  renderActions() {
    const { undoChanges, saveChanges, isMobile } = this.props
    const cssButton = classNames('border-btn', { 'icon-btn': isMobile, 'remove-btn': isMobile })
    return (
      <ul className="changes-btns">
        <li>
          <button
            data-header-cancel="true"
            className={cssButton}
            onClick={undoChanges}
          >
            Cancel
          </button>
        </li>
        <li>
          <button
            data-header-save="true"
            className="border-btn"
            onClick={saveChanges}
          >
            Save Changes
          </button>
        </li>
      </ul>
    )
  }

  render() {
    const { canEdit, editMode, loading } = this.props

    if (loading) {
      return this.renderLoading()
    }

    if (editMode && canEdit) {
      return this.renderActions()
    }

    if (canEdit) {
      return this.renderEditAction()
    }

    return null
  }
}

Actions.defaultProps = {
  loading: false,
  canEdit: false,
  editMode: false
}

Actions.propTypes = {
  isMobile: PropTypes.bool.isRequired,
  loading: PropTypes.bool,
  canEdit: PropTypes.bool,
  editLabel: PropTypes.string.isRequired,
  editMode: PropTypes.bool,
  toggleEditMode: PropTypes.func.isRequired,
  saveChanges: PropTypes.func.isRequired,
  undoChanges: PropTypes.func.isRequired
}

export default Actions
