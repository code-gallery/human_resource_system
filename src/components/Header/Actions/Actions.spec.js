import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import Actions from './Actions'

describe('components/Header/Actions', () => {
  const commonProps = {
    isMobile: false,
    editLabel: 'Edit',
    toggleEditMode: jest.fn(),
    saveChanges: jest.fn(),
    undoChanges: jest.fn()
  }

  describe('with defaultProps', () => {
    it('renders correctly', () => {
      const props = commonProps

      const tree = renderer.create(
        <Actions {...props} />
      ).toJSON()

      expect(tree).toMatchSnapshot()
    })
  })

  describe('when loading is true', () => {
    it('renders correctly', () => {
      const props = {
        ...commonProps,
        loading: true
      }

      const tree = renderer.create(
        <Actions {...props} />
      ).toJSON()

      expect(tree).toMatchSnapshot()
    })
  })

  describe('when editMode and canEdit are true', () => {
    let wrapper
    const props = {
      ...commonProps,
      editMode: true,
      canEdit: true
    }

    it('renders correctly', () => {
      const tree = renderer.create(
        <Actions {...props} />
      ).toJSON()

      expect(tree).toMatchSnapshot()
    })

    it('calls props.undoChanges correctly', () => {
      wrapper = shallow(
        <Actions {...props} />
      )

      expect(props.undoChanges).not.toHaveBeenCalled()
      wrapper.find('.changes-btns [data-header-cancel]').simulate('click')
      expect(props.undoChanges).toHaveBeenCalled()
    })

    it('calls props.saveChanges correctly', () => {
      wrapper = shallow(
        <Actions {...props} />
      )

      expect(props.saveChanges).not.toHaveBeenCalled()
      wrapper.find('.changes-btns [data-header-save]').simulate('click')
      expect(props.saveChanges).toHaveBeenCalled()
    })
  })

  describe('when canEdit is true', () => {
    let wrapper
    let props
    it('renders correctly', () => {
      props = {
        ...commonProps,
        canEdit: true
      }

      const tree = renderer.create(
        <Actions {...props} />
      ).toJSON()

      expect(tree).toMatchSnapshot()
    })

    it('calls props.toggleEditMode correctly', () => {
      wrapper = shallow(
        <Actions {...props} />
      )

      expect(props.toggleEditMode).not.toHaveBeenCalled()
      wrapper.find('.changes-btns .border-btn').simulate('click')
      expect(props.toggleEditMode).toHaveBeenCalled()
    })
  })
})
