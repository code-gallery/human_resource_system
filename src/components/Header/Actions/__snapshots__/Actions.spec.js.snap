// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Header/Actions when canEdit is true renders correctly 1`] = `
<ul
  className="changes-btns"
>
  <li>
    <button
      className="border-btn"
      onClick={[Function]}
    >
      Edit
    </button>
  </li>
</ul>
`;

exports[`components/Header/Actions when editMode and canEdit are true renders correctly 1`] = `
<ul
  className="changes-btns"
>
  <li>
    <button
      className="border-btn"
      data-header-cancel="true"
      onClick={[Function]}
    >
      Cancel
    </button>
  </li>
  <li>
    <button
      className="border-btn"
      data-header-save="true"
      onClick={[Function]}
    >
      Save Changes
    </button>
  </li>
</ul>
`;

exports[`components/Header/Actions when loading is true renders correctly 1`] = `
<ul
  className="changes-btns"
>
  <li>
    <img
      alt="loading"
      src="/assets/loading.gif"
    />
  </li>
</ul>
`;

exports[`components/Header/Actions with defaultProps renders correctly 1`] = `null`;
